### DISCUSSION TITLE
Finding markers of all cluster using a loop

### DISCUSSION THREAD

**Discussion by arturomarin** [satijalab/seurat#4027] [Q&A] [answered]
Hi, 
I am new with Seurat and with R, but I have some programming skill. I am try to do the identification of conserved cell type markers for all the clusters using a for loop, but it doesn't work. The loop is:

```
for (i in AllClus) {
DefaultAssay(immune.combined) <- "RNA"
gmarker <- paste("g", i, sep = "")
gmarker.markers <- FindConservedMarkers(immune.combined, ident.1 = i, grouping.var = "Condition", verbose = FALSE)
}
```

where "AllClus" is an array that have the number of all the cluster in the data (0, 1, 2 ....). But, when I use this code the variable g0.markers, g1.markers ... don't exist. Only exist gmarker.markers variable. Someone can help me?

**Comment by timoast** [comment]
`gmarker` is not used here, and `gmarker.markers` is overwritten at each iteration of the loop. I think you might have intended to append the markers to a list, eg:

```r
DefaultAssay(immune.combined) <- "RNA"
all_markers <- list()
for (i in AllClus) {
    gmarker <- paste("g", i, sep = "")
    all_markers[[gmarker]] <- FindConservedMarkers(immune.combined, ident.1 = i, grouping.var = "Condition", verbose = FALSE)
}
```

**Reply by arturomarin** [reply]
Hi,

it seems this do not work. Before your code I did:

AllClus <- c(0,1)

And when I use your code the markers of the g0 cluster do not exist:

```
> gmarkers[0]
character(0)
> gmarkers[1]
[1] "g1"
> gmarkers[2]
[1] NA

```

Any suggestion?

**Comment by Pancreas-Pratik** [comment, marked as answer]
Hi @arturomarin,

I do something similar like this all the time.

Try this:

```
for (i in AllClus) {
  DefaultAssay(immune.combined) <- "RNA"
  gmarker.markers <- FindConservedMarkers(immune.combined, ident.1 = i, grouping.var = "Condition", verbose = FALSE)
  assign(paste0("g", i, ".markers"), gmarker.markers)
  }
#cleanup environment
rm("gmarker.markers", "AllClus")
```

Hope this helps!

**Comment by arturomarin** [comment]
Hi @Pancreas-Pratik your script works. Thank you very much!!
